README.md
# File_Search
这是用pyqt5开发的文件搜索工具，可帮助您快速、高效检索并预览电脑上的文件。它不仅包含文件搜索的基本功能，还内置了图片、文本文档、视频、音频预览器，您可快速知悉该文件所含内容。
# 特点
1. 文件搜索极致优化，占用尽可能少的资源，实现高效检索
2. 可预览常见的视频、图片、文档、音频格式文件，且内置了所需的解码器，不需要额外依赖，免安装运行，真正实现高兼容性，高稳定性，高便捷性
3. 功能齐全完整，含有常见文件搜索工具的基本功能
4. 自动储存搜索历史记录，可设置记录储存条数
5. 界面美观，部件大小合理，预览区域可拉伸与隐藏。


changelog.txt
File_Search  1.0 版本：
  发布于 2022/3/15

- 具备基本的文件搜索功能
- 可保存搜索关键字历史记录

videoplayer.py
from PyQt5 import QtGui, QtCore
from PyQt5.QtCore import Qt, QTimer
from PyQt5.QtWidgets import (QApplication, QHBoxLayout, QLabel, QPushButton, QSizePolicy, QSlider,
                             QStyle, QVBoxLayout, QWidget)
import time
from ffpyplayer.player import MediaPlayer


class VideoPlayer(QWidget):

    def __init__(self, video_path, parent=None):
        super(VideoPlayer, self).__init__(parent)

        self.video_path = video_path

        self.setWindowTitle("视频预览")

        self.label = QLabel()
        self.label.setScaledContents(True)
        # self.label.resize(self.width(), self.height())

        # 重要代码，实现窗口可缩放
        # https://stackoverflow.com/questions/24940709/qt-designer-window-wont-get-smaller-than-a-qlabel-with-pixmap
        self.label.setMinimumSize(1, 1)

        self.playButton = QPushButton()
        self.playButton.setIcon(self.style().standardIcon(QStyle.SP_MediaPlay))
        self.playButton.clicked.connect(self.play)

        self.pauseButton = QPushButton()
        self.pauseButton.setIcon(self.style().standardIcon(QStyle.SP_MediaPause))
        self.pauseButton.clicked.connect(self.pause)
        self.pauseButton.hide()

        self.positionSlider = QSlider(Qt.Horizontal)
        self.positionSlider.setRange(0, 0)
        self.positionSlider.sliderMoved.connect(self.setPosition)

        self.durationLabel = QLabel()
        self.durationLabel.setSizePolicy(QSizePolicy.Preferred,
                QSizePolicy.Maximum)

        controlLayout = QHBoxLayout()
        controlLayout.setContentsMargins(0, 0, 0, 0)
        controlLayout.addWidget(self.playButton)
        controlLayout.addWidget(self.pauseButton)
        controlLayout.addWidget(self.positionSlider)
        controlLayout.addWidget(self.durationLabel)

        layout = QVBoxLayout()
        layout.addWidget(self.label)
        layout.addLayout(controlLayout)

        self.setLayout(layout)

        self.timer = QTimer()
        self.timer.timeout.connect(self.showFrame)
        self.timer.setInterval(1000 / 8)   # 控制帧数

        self.video_media_data = None
        self.init_player()

    def init_player(self):
        """
        初始化播放器，获取第一帧画面及视频总时长
        """
        self.player = MediaPlayer(self.video_path)
        while 1:
            if self.showFrame()[0] is not None:
                self.video_media_data = self.player.get_metadata()
                self.player.set_pause(True)
                break
        self.positionSlider.setRange(0, self.video_media_data['duration'])   # 设置进度条长度
        self.durationLabel.setText(f"0/{self.video_media_data['duration']}")

    def play(self):
        if not self.timer.isActive():
            self.player.set_pause(False)
            self.timer.start()
            self.positionSlider.setValue(0)
        else:
            self.player.set_pause(False)
        self.playButton.hide()
        self.pauseButton.show()

    def pause(self):
        self.player.set_pause(True)
        self.pauseButton.hide()
        self.playButton.show()

    def showFrame(self):
        frame, val = self.player.get_frame()

        if val != 'eof' and frame is not None:
            time.sleep(val)
            img, t = frame

            data = img.to_bytearray()[0]
            width, height = img.get_size()

            # the technical name for the 'rgb24' default pixel format is RGB888,
            # which is QImage.Format_RGB888 in the QImage format enum
            qimage = QtGui.QImage(data, width, height, QtGui.QImage.Format_RGB888)
            self.pixmap = QtGui.QPixmap.fromImage(qimage)
            self.pixmap = self.pixmap.scaled(self.label.width(), self.label.height(),
                                   QtCore.Qt.IgnoreAspectRatio)
            self.label.setPixmap(self.pixmap)
            self.update()

            self.positionSlider.setValue(frame[1])
            if self.video_media_data is not None:
                self.durationLabel.setText(f"{round(frame[1], 1)}/{self.video_media_data['duration']}")

        elif val == "eof":
            self.player.seek(0, relative=False, accurate=False)            # 跳转回视频开头
            self.player.set_pause(True)
            self.pauseButton.hide()
            self.playButton.show()
            self.timer.stop()
        elif frame is None:
            time.sleep(0.01)
        return frame, val

    def setPosition(self, position):
        pass

    def closeEvent(self, event):
        self.timer.stop()
        self.player.close_player()


if __name__ == '__main__':

    import sys

    app = QApplication(sys.argv)

    player = VideoPlayer()
    player.resize(320, 240)
    player.show()

    sys.exit(app.exec_())

preview_area.py
from ffpyplayer import tools
from PyQt5.QtGui import QIcon, QFont, QPixmap, QPainter, QImage, QPalette, QImageReader
from PyQt5.QtCore import Qt
from PyQt5.QtPrintSupport import QPrintDialog, QPrinter
from PyQt5.QtWidgets import (QApplication, QFileDialog, QHBoxLayout, QWidget, QLabel, QMainWindow, QAction, QMenu,
                             QScrollArea, QSizePolicy, QTextEdit)
from videoplayer import VideoPlayer


class PreviewArea(QWidget):
    def __init__(self):
        super(PreviewArea, self).__init__()

        self.support_formats = {}

        self.get_support_formats()

        self.video_player = None
        self.image_viewer = ImageViewer()
        self.image_viewer.hide()
        self.text_viewer = QTextEdit()
        self.text_viewer.setReadOnly(True)
        self.text_viewer.hide()

        self.hboxLayout = QHBoxLayout()
        self.hboxLayout.addWidget(self.image_viewer)
        self.hboxLayout.addWidget(self.text_viewer)
        self.setLayout(self.hboxLayout)

    def show_image(self, image_path):
        self.image_viewer.open(image_path)
        self.text_viewer.hide()
        self.show()
        self.image_viewer.show()

    def open_video(self, video_path):
        self.video_player = VideoPlayer(video_path)
        self.show()
        self.video_player.show()

    def show_text(self, text_file_path):
        file_coding = self.get_file_encoding(text_file_path)
        if file_coding == "unknown":
            return "unknown code"
        with open(text_file_path, "r", encoding=file_coding) as f:
            content = f.read()

        self.text_viewer.setText(content)
        self.show()
        self.image_viewer.hide()
        self.text_viewer.show()

    def get_support_formats(self):
        self.support_formats["image"] = []
        self.support_formats["video"] = []
        for i in QImageReader.supportedImageFormats():
            self.support_formats["image"].append(bytes(i).decode())
        video_support_formats = tools.get_fmts(False, True)[2]
        self.support_formats["video"] = [x for x in video_support_formats if x]    # 去除列表空白元素
        self.support_formats["text"] = ["txt", "ini", "doc", "docx"]

    def get_file_encoding(self, file_path):
        """
        获取字符编码类型
        参考: https://blog.csdn.net/u013314786/article/details/77931548
        :param file_path: 要辨别的文件路径
        """

        codes = ['UTF-8', 'GB18030', 'BIG5', "gbk", "UTF-16"]
        # UTF-8 BOM前缀字节
        utf8_bom = b'\xef\xbb\xbf'
        # 遍历编码类型

        with open(file_path, 'rb') as f:
            data = f.read()
        for code in codes:
            try:
                data.decode(encoding=code)
                if 'UTF-8' == code and data.startswith(utf8_bom):
                    return 'UTF-8-SIG'
                return code
            except UnicodeDecodeError:
                continue
        return 'unknown'


class ImageViewer(QMainWindow):
    def __init__(self):
        super(ImageViewer, self).__init__()

        self.printer = QPrinter()
        self.scaleFactor = 0.0

        self.imageLabel = QLabel()
        self.imageLabel.setBackgroundRole(QPalette.Base)
        self.imageLabel.setSizePolicy(QSizePolicy.Minimum, QSizePolicy.Minimum)
        self.imageLabel.setScaledContents(True)

        self.scrollArea = QScrollArea()
        self.scrollArea.setBackgroundRole(QPalette.Dark)
        self.scrollArea.setWidget(self.imageLabel)
        self.setCentralWidget(self.scrollArea)

        self.createActions()
        self.createMenu()

        self.setWindowTitle("Image Viewer")

    def open(self, image_path):
        if image_path:
            image = QImage(image_path)
            self.imageLabel.setPixmap(QPixmap.fromImage(image))
            self.scaleFactor = 1.0

            self.fitToWindowAct.setEnabled(True)
            self.updateActions()

            if not self.fitToWindowAct.isChecked():
                self.imageLabel.adjustSize()

    def print_(self):
        dialog = QPrintDialog(self.printer, self)
        if dialog.exec_():
            painter = QPainter(self.printer)
            rect = painter.viewport()
            size = self.imageLabel.pixmap().size()
            size.scale(rect.size(), Qt.KeepAspectRatio)
            painter.setViewport(rect.x(), rect.y(), size.width(), size.height())
            painter.setWindow(self.imageLabel.pixmap().rect())
            painter.drawPixmap(0, 0, self.imageLabel.pixmap())

    def zoomIn(self):
        self.scaleImage(1.25)

    def zoomOut(self):
        self.scaleImage(0.8)

    def normalSize(self):
        self.imageLabel.adjustSize()
        self.scaleFactor = 1.0

    def fitToWindow(self):
        fitToWindow = self.fitToWindowAct.isChecked()
        self.scrollArea.setWidgetResizable(fitToWindow)
        if not fitToWindow:
            self.normalSize()

        self.updateActions()

    def createActions(self):
        self.zoomInAct = QAction("Zoom &In (25%)", self, shortcut="Ctrl++",
                enabled=False, triggered=self.zoomIn)

        self.zoomOutAct = QAction("Zoom &Out (25%)", self, shortcut="Ctrl+-",
                enabled=False, triggered=self.zoomOut)

        self.normalSizeAct = QAction("&Normal Size", self, shortcut="Ctrl+S",
                enabled=False, triggered=self.normalSize)

        self.fitToWindowAct = QAction("&Fit to Window", self, enabled=False,
                checkable=True, shortcut="Ctrl+F", triggered=self.fitToWindow)


    def createMenu(self):
        self.viewMenu = QMenu("&View", self)
        self.viewMenu.addAction(self.zoomInAct)
        self.viewMenu.addAction(self.zoomOutAct)
        self.viewMenu.addAction(self.normalSizeAct)
        self.viewMenu.addSeparator()
        self.viewMenu.addAction(self.fitToWindowAct)

        self.menuBar().addMenu(self.viewMenu)

    def updateActions(self):
        self.zoomInAct.setEnabled(not self.fitToWindowAct.isChecked())
        self.zoomOutAct.setEnabled(not self.fitToWindowAct.isChecked())
        self.normalSizeAct.setEnabled(not self.fitToWindowAct.isChecked())

    def scaleImage(self, factor):
        self.scaleFactor *= factor
        self.imageLabel.resize(self.scaleFactor * self.imageLabel.pixmap().size())

        self.adjustScrollBar(self.scrollArea.horizontalScrollBar(), factor)
        self.adjustScrollBar(self.scrollArea.verticalScrollBar(), factor)

        self.zoomInAct.setEnabled(self.scaleFactor < 3.0)
        self.zoomOutAct.setEnabled(self.scaleFactor > 0.333)

    def adjustScrollBar(self, scrollBar, factor):
        scrollBar.setValue(int(factor * scrollBar.value()
                                + ((factor - 1) * scrollBar.pageStep()/2)))




main.py
import sys
import os
from PyQt5 import QtGui, QtWidgets, Qt, QtCore
from PyQt5.QtWidgets import QTableWidget, QTableWidgetItem, QMessageBox, QComboBox, QSizePolicy, QSplitter, QAbstractItemView
import time
import pickle
import ctypes
import resource
import subprocess
from functools import partial
from preview_area import PreviewArea


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self, parent):
        super().__init__(parent=parent)

        self.setWindowTitle("file search")
        self.resize(1000, 800)

        self.search_frame = QtWidgets.QLineEdit()
        self.search_frame.setAlignment(QtCore.Qt.AlignLeft)

        self.search_button = QtWidgets.QPushButton("搜索")
        self.search_button.clicked.connect(self.click_search_button)

        self.choose_search_dir_button = QtWidgets.QPushButton("选择路径")
        self.choose_search_dir_button.clicked.connect(self.choose_search_dir)

        self.combobox = QComboBox(self)
        self.combobox.setLineEdit(self.search_frame)
        self.combobox.setMinimumWidth(self.width()-60)
        self.combobox.setMinimumHeight(26)
        self.combobox.setCompleter(None)
        self.combobox.activated.connect(self.choose_history)

        view = QtWidgets.QListView(self.combobox)
        view.setMinimumHeight(420)

        self.combobox.setView(view)

        hbox_layout = QtWidgets.QHBoxLayout()
        hbox_layout.addWidget(self.combobox, stretch=0, alignment=QtCore.Qt.AlignRight)
        hbox_layout.addWidget(self.choose_search_dir_button, alignment=QtCore.Qt.AlignRight)
        hbox_layout.addWidget(self.search_button, stretch=0, alignment=QtCore.Qt.AlignRight)
        hbox_layout.addStretch(10)

        # 初始化table_widget
        self.table_widget = ShowResultsTable()
        self.table_widget.setColumnCount(4)
        self.table_widget.setHorizontalHeaderLabels(["文件名", "路径", '创建时间', "修改时间"])
        self.table_widget.doubleClicked.connect(self.table_widget.open_click_file)
        self.table_widget.cellClicked.connect(self.preview_table_cell)
        # 右键菜单策略
        # self.table_widget.setContextMenuPolicy(QtCore.Qt.ContextMenuPolicy.DefaultContextMenu)
        # 下面的方法对于调整大小很重要
        self.table_widget.horizontalHeader().resizeSections(QtWidgets.QHeaderView.Stretch)

        self.preview_area = PreviewArea()
        self.preview_area.hide()         # 默认隐藏

        self.splitter = QSplitter()

        self.splitter.addWidget(self.table_widget)
        self.splitter.addWidget(self.preview_area)

        vbox_layout = QtWidgets.QVBoxLayout()
        vbox_layout.setSpacing(15)
        vbox_layout.addLayout(hbox_layout)
        vbox_layout.addWidget(self.splitter)

        widget = QtWidgets.QWidget()
        widget.setLayout(vbox_layout)

        self.statusBar().showMessage("0个对象")

        self.setCentralWidget(widget)
        self.history = History(self.combobox)
        self.search_path = None
        self.file_search_thread = None

    def click_search_button(self):
        self.search_button.setEnabled(False)
        self.search_button.setText("正在搜索")
        self.history.save_history(self.search_frame.text())
        file_name = self.search_frame.text()
        if file_name == "":
            QMessageBox.warning(self, "警告", "你还没有输入文件名", buttons=QMessageBox.Ok)
        else:
            self.threadpool = QtCore.QThreadPool()
            self.file_search_thread = FileSearchThread(file_name, self.search_path)
            self.threadpool.start(self.file_search_thread)
            self.file_search_thread.signals.finished.connect(self.file_search_finish)

    def choose_search_dir(self):
        dir_choose = QtWidgets.QFileDialog.getExistingDirectory(self, "选取搜索文件夹")
        if dir_choose != "":
            self.search_path = dir_choose

    def file_search_finish(self):
        self.table_widget.clearContents()
        self.table_widget.setRowCount(0)
        if len(self.file_search_thread.result) > self.table_widget.rowCount():
            self.table_widget.setRowCount(len(self.file_search_thread.result))
        for i, info_list in enumerate(self.file_search_thread.result):
            for j, info in enumerate(info_list):
                table_cell = QTableWidgetItem(info)
                table_cell.setToolTip(info)
                if j == 0:
                    file_icon = QtWidgets.QFileIconProvider().icon(QtCore.QFileInfo(info_list[1]))
                    table_cell.setIcon(file_icon)
                self.table_widget.setItem(i, j, table_cell)
        self.statusBar().showMessage(f"{len(self.file_search_thread.result)}个对象")
        self.search_button.setText("搜索")
        self.table_widget.horizontalHeader().resizeSections(QtWidgets.QHeaderView.Stretch)
        self.search_button.setEnabled(True)
        self.file_search_thread.is_running = False

    def choose_history(self, event):
        choose_text = self.combobox.itemText(event)
        self.search_frame.setText(choose_text)

    def closeEvent(self, event):
        event.accept()
        if self.file_search_thread is not None:
            if self.file_search_thread.is_running:
                self.file_search_thread.is_running = False

    def preview_table_cell(self, row, col):
        if col == 0:
            try:
                file_name = self.table_widget.item(row, col).text()
            except Exception:
                pass
            else:
                file_name_extension = file_name.split(".")[-1]
                if file_name_extension in self.preview_area.support_formats["text"]:
                    self.preview_area.show_text(self.table_widget.item(row, 1).text())
                elif file_name_extension in self.preview_area.support_formats["image"]:
                    self.preview_area.show_image(self.table_widget.item(row, 1).text())
                elif file_name_extension == self.preview_area.support_formats["video"]:
                    self.preview_area.open_video(self.table_widget.item(row, 1).text())


class ShowResultsTable(QTableWidget):
    def __init__(self):
        super(ShowResultsTable, self).__init__()

        self.setShowGrid(False)
        self.verticalHeader().setVisible(False)
        self.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Interactive)
        self.horizontalHeader().setStretchLastSection(True)
        self.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)
        self.setSelectionMode(QAbstractItemView.NoSelection)
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)

    def open_click_file(self, index):
        row = index.row()
        col = index.column()
        if col == 0:
            try:
                file_path = self.item(row, 1).text()
                os.startfile(file_path)
            except Exception:
                pass

    def open_folder(self, index):
        row = index.row()
        try:
            file_path = self.item(row, 1).text()
            cmd = 'explorer.exe /select, "{}", vbNormalFocus'.format(file_path)
            si = subprocess.STARTUPINFO()
            si.wShowWindow = subprocess.SW_HIDE
            subprocess.call(cmd, startupinfo=si)
        except Exception:
            pass

    def contextMenuEvent(self, event):
        index = self.currentIndex()
        menu = QtWidgets.QMenu(self)

        default_open_file = QtWidgets.QAction("使用默认应用打开文件")
        default_open_file.triggered.connect(partial(self.open_click_file, index))
        open_dir = QtWidgets.QAction("打开文件目录")
        open_dir.triggered.connect(partial(self.open_folder, index))

        menu.addAction(default_open_file)
        menu.addAction(open_dir)
        menu.exec(event.globalPos())


class FileSearchThread(QtCore.QRunnable):
    def __init__(self, file_name, search_path):
        super(FileSearchThread, self).__init__()
        self.file_name = file_name
        self.search_path = search_path
        self.result = None
        self.is_running = True
        self.signals = Signals()

    @QtCore.pyqtSlot()
    def run(self):
        self.result = self.search_file(self.file_name, self.search_path)
        self.signals.finished.emit()

    def search_file(self, name, path):
        result = []
        if path is None:
            path = "/"
        for root, folder, files in os.walk(path):
            if not self.is_running:
                break
            for file_name in files:
                if name in file_name:
                    if not self.is_running:
                        break
                    file_path = os.path.abspath(os.path.join(root, file_name))

                    stat_info = os.stat(file_path)

                    # 时间戳转换
                    create_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(stat_info.st_ctime))
                    modif_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(stat_info.st_mtime))

                    result.append([file_name, file_path, create_time, modif_time])  # 保存路径与盘符
        return result


class History(object):
    def __init__(self, combobox):
        self.history_number = None
        self.max_history_number = 20
        appdata_path = os.getenv("APPDATA")
        self.app_dir_path = os.path.join(appdata_path, "file_search")
        self.history_file_path = os.path.join(self.app_dir_path, "history.pickle")
        self.combobox = combobox
        self.init_combobox()

    def init_combobox(self):
        history = self.read_history()
        if history is not None:
            self.combobox.addItems(history)

    def save_history(self, text):
        current_history = self.read_history()
        if current_history is None:
            current_history = [text]
        else:
            if self.history_number == self.max_history_number:
                current_history.pop(-1)
            current_history.insert(0, text)
        self.combobox.clear()
        self.combobox.addItems(current_history)
        with open(self.history_file_path, "wb") as f:
            pickle.dump(current_history, f)

    def read_history(self):
        if not os.path.isdir(self.app_dir_path):
            os.mkdir(self.app_dir_path)
        if not os.path.isfile(self.history_file_path):
            with open(self.history_file_path, "wb+") as f:
                pass
            self.history_number = 0
            return None
        else:
            with open(self.history_file_path, "rb") as f:
                try:
                    file_data = pickle.load(f)
                    if file_data is not None:
                        self.history_number = len(file_data)
                except EOFError:
                    return None
            return file_data


class Signals(QtCore.QObject):
    finished = QtCore.pyqtSignal()


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    ctypes.windll.shell32.SetCurrentProcessExplicitAppUserModelID("alairack")
    app.setWindowIcon(QtGui.QIcon(":/icons/logo.ico"))
    main_window = MainWindow(None)

    main_window.show()
    sys.exit(app.exec_())
